cpp/include/raft/core/mdbuffer.cuh

```diff
@@ -0,0 +1,831 @@
+            return view_type_variant{inner};
+          } else {
+            return view_type_variant{inner.view()};
+          }
+        },
+        data_);
+    }
+  }
+
+  template <typename MemTypeConstant>
+  [[nodiscard]] auto view() const
+  {
+    if constexpr (MemTypeConstant::value.has_value()) {
+      if (is_owning()) {
+        return make_const_mdspan(
+          std::get<owning_type<MemTypeConstant::value.value()>>(data_).view());
+      } else {
+        return make_const_mdspan(std::get<view_type<MemTypeConstant::value.value()>>(data_));
+      }
+    } else {
+      return std::visit(
+        [](auto&& inner) {
+          if constexpr (is_mdspan_v<std::remove_reference_t<decltype(inner)>>) {
+            return const_view_type_variant{make_const_mdspan(inner)};
+          } else {
+            return const_view_type_variant{make_const_mdspan(inner.view())};
+          }
+        },
+        data_);
+    }
+  }
+
+ public:
+  /**
+   * @brief Return an mdspan of the indicated memory type representing a view
+   * on the stored data. If the mdbuffer does not contain data of the indicated
+   * memory type, a std::bad_variant_access will be thrown.
+   */
+  template <memory_type mem_type>
+  [[nodiscard]] auto view()
+  {
+    return view<memory_type_constant<mem_type>>();
+  }
+  /**
+   * @brief Return an mdspan containing const elementgs of the indicated memory type representing a
+   * view on the stored data. If the mdbuffer does not contain data of the indicated memory type, a
+   * std::bad_variant_access will be thrown.
+   */
+  template <memory_type mem_type>
+  [[nodiscard]] auto view() const
+  {
+    return view<memory_type_constant<mem_type>>();
+  }
+  /**
+   * @brief Return a std::variant representing the possible mdspan types that
+   * could be returned as views on the mdbuffer. The variant will contain the mdspan
+   * corresponding to its current memory type.
+   *
+   * This method is useful for writing generic code to handle any memory type
+   * that might be contained in an mdbuffer at a particular point in a
+   * workflow. By performing a `std::visit` on the returned value, the caller
+   * can easily dispatch to the correct code path for the memory type.
+   */
+  [[nodiscard]] auto view() { return view<memory_type_constant<>>(); }
+  /**
+   * @brief Return a std::variant representing the possible mdspan types that
+   * could be returned as const views on the mdbuffer. The variant will contain the mdspan
+   * corresponding to its current memory type.
+   *
+   * This method is useful for writing generic code to handle any memory type
+   * that might be contained in an mdbuffer at a particular point in a
+   * workflow. By performing a `std::visit` on the returned value, the caller
+   * can easily dispatch to the correct code path for the memory type.
+   */
+  [[nodiscard]] auto view() const { return view<memory_type_constant<>>(); }
+};
+
```

Comment:
This is just a suggestion and perhaps others can weigh in. I was thinking if there should be a helper to access the underlying mdarray from the mdbuffer object. However, I cannot think of a use-case where this will be helpful.

On the other hand if we want the mdbuffer to act like a fundamental container (wherein the fact that the underlying container is an mdarray is unknown to the user), the `data_handle` method can be useful to fetch the ptr to the underlying data (rather than the user going through the view using `.view().data_handle()`

Reply:
I'd prefer to avoid exposing the underlying `mdarray` if at all possible. It won't always exist, and if the caller wants to directly manipulate the `mdarray`, I think that suggests that they should not be using an `mdbuffer` in the first place. The general purpose of `mdbuffer` is to get you an `mdspan` of the correct type when you have an `mdspan` of the possibly-correct-but-maybe-incorrect type.

In terms of exposing `data_handle`, I'm less strongly opposed, but I would prefer not to. Basically, the goal is to encourage folks to write `mdspan` based code wherever possible. `mdbuffer` is used to translate between mdspans, but it should not be used to manipulate the underlying data directly.